Added a script to perform " Validation of constant flushing of data when SSD write back cache feature is enabled"

preethisn opened the pull request:
Signed-off-by: Preethi Nataraj <pnataraj@localhost.localdomain>

# Description validate constant flushing and evicting of data for an extended period of time.

preethisn commented:
logs copied to http://magna002.ceph.redhat.com/ceph-qe-logs/Preethi/ssd.txt

psathyan commented on utility/SSD_writeback_cache.py:
```diff
@@ -0,0 +1,64 @@
+
+    """
+    size = 1 * 1024 * 3
+    pool_name = "ssd"
+    image_name = "image3"
```
Would it make sense to retrieve these values from CLI? 

The current implementation creates the required resources but does not cleanup after the workflow completion. Can we implement a teardown sequence?

sunilkumarn417 replied:
https://github.com/red-hat-storage/cephci/blob/master/pipeline/scripts/ci/cos_cli.py please refer this for custom values which should be picked from CLI.

preethisn replied:
Yes, implemented via cli with custom values as per the reference provided

psathyan commented on utility/SSD_writeback_cache.py:
```diff
@@ -0,0 +1,64 @@
+    size = 1 * 1024 * 3
+    pool_name = "ssd"
+    image_name = "image3"
+    os.system(f"ceph osd pool create {pool_name} 128")
+    os.system(f"ceph osd pool application enable {pool_name} rbd")
+    os.system(f"rbd pool init -p {pool_name}")
+    os.system(f"rbd create --size {size} {pool_name}/{image_name} --thick-provision")
```
These resources are required for testing... please verify that they are created successfully before proceeding to the next step.

preethisn replied:
Yes, these are working fine and captured in logging.

psathyan commented on utility/SSD_writeback_cache.py:
```diff
@@ -0,0 +1,64 @@
+    os.system(f"rbd pool init -p {pool_name}")
+    os.system(f"rbd create --size {size} {pool_name}/{image_name} --thick-provision")
+    for i in range(1, 10000):
+        print("#" * 80)
```
IMO, we can avoid this line.... we would have not many

psathyan commented on utility/SSD_writeback_cache.py:
```diff
@@ -0,0 +1,64 @@
+    os.system(f"ceph osd pool application enable {pool_name} rbd")
+    os.system(f"rbd pool init -p {pool_name}")
+    os.system(f"rbd create --size {size} {pool_name}/{image_name} --thick-provision")
+    for i in range(1, 10000):
```
Can this be made as a configurable?

preethisn replied:
i dont think so we can make it as configurable

psathyan commented on utility/SSD_writeback_cache.py:
```diff
@@ -0,0 +1,64 @@
+            ],
+            shell=True,
+        )
+        timeout = random.randint(45, 100)
```
IMO, we should check the storage capacity before going to the next iteration. We could break the cluster if it is not properly sized.

preethisn replied:
As discussed, will consider this storage capacity as improvements and enhancement in next sprint

psathyan commented on utility/SSD_writeback_cache.py:
```diff
@@ -0,0 +1,64 @@
+            shell=True,
+        )
+        timeout = random.randint(45, 100)
+        print(f"sleeping for {timeout}")
```
opt to use `log` instead of `print`

preethisn replied:
done

psathyan commented on utility/SSD_writeback_cache.py:
```diff
@@ -0,0 +1,64 @@
+
+
+if __name__ == "__main__":
+    rc = ssd_write_cache()
```
There is no return value for the method `ssd_write_cache`

preethisn commented:
New logs attached 
[logs_ssd.txt](https://github.com/red-hat-storage/cephci/files/8178998/logs_ssd.txt)

psathyan reviewed (commented):
minor changes...

preethisn commented:
Changes are done. Please find the latest logs.
[logs_ssd.txt](https://github.com/red-hat-storage/cephci/files/8184474/logs_ssd.txt)

sunilkumarn417 commented on utility/ssd_writeback_cache.py:
```diff
@@ -0,0 +1,95 @@
+This script helps to validate the write back cache flushing of data
+
+usage:
+SSD_writeback_cache.py  --pool POOL --image IMAGE [--size=<size>] [--iteration=<iteration>]
```
please replace module name here.

preethisn replied:
done

sunilkumarn417 commented on utility/ssd_writeback_cache.py:
```diff
@@ -0,0 +1,95 @@
+"""
+
+
+def ssd_write_cache(args):
```
I think we might not need parameter `args`.

preethisn replied:
removed and tested. its working fine

sunilkumarn417 commented on utility/ssd_writeback_cache.py:
```diff
@@ -0,0 +1,95 @@
+    iteration = int(args["--iteration"])
+    pool_name = args["--pool"]
+    image_name = args["--image"]
+    ssd_write_cache(args)
```
please include try-finally clause to have automatic teardown of RBD pool.

preethisn replied:
Thanks Sunil. We take this input in our next PR

preethisn commented on utility/ssd_writeback_cache.py:
```diff
@@ -0,0 +1,100 @@
+    args = docopt(doc)
+    image_size = 1024
+    if args.get("--size"):
+        image_size = int(args["<size>"]) * image_size
```
This change is implemented as suggested.

preethisn commented:
@sunilkumarn417 @psathyan I have addressed the review comments except below and we consider them in our next PR. 
- please include try-finally clause to have automatic teardown of RBD pool.
- calculate storage capacity before moving to next iteration
Please take this PR to closure if no concerns.

psathyan reviewed (commented):
This looks good as the plan is to take the improvements later

sunilkumarn417 reviewed (approved):
As improvements is considered in next PR, let's get this code in.
